### PR TITLE
Torothin high tide efficiency for Issue #2631

### PR DESCRIPTION
**Pull request by torothin**

### CODE REVIEW

File: src/parser/shaman/restoration/modules/talents/HighTide.js [hunk 1]
```diff
@@ -21,22 +21,65 @@ const bounceReduction = 0.7;
...
+    // these are for tracking high tide efficiency
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(SPELLS.CHAIN_HEAL), this.onChainHealCast);
+    this.addEventListener(Events.applybuff, this.onHighTideBuff);
+    this.addEventListener(Events.applybuffstack, this.onHighTideBuff); //possible to get a re-application of the buff
+    this.addEventListener(Events.removebuff, this.onHighTideRemoveBuff);
+  }
+
+  // in the event of a High Tide buff, if a buff still exists adds it to unused then resets buff counter to 2
+  onHighTideBuff(event) {
+    if(event.ability.guid === 288675) {
```
Comment (niseko) [inline src/parser/shaman/restoration/modules/talents/HighTide.js]:
```suggestion
    if(event.ability.guid === SPELLS.HIGH_TIDE_BUFF.id) {
```

File: src/parser/shaman/restoration/modules/talents/HighTide.js [hunk 2]
```diff
@@ -21,22 +21,65 @@ const bounceReduction = 0.7;
...
+      this.unusedHighTides += this.currentHighTideBuff; //not sure if this is possible
+      this.currentHighTideBuff = 2;
+    }
+  }
+
+  // in the event of a High Tide remove buff, adds any existing buff to unused and sets buff counter to 0
+  onHighTideRemoveBuff(event) {
+    if(event.ability.guid === 288675) {
```
Comment (niseko) [inline src/parser/shaman/restoration/modules/talents/HighTide.js]:
```suggestion
    if(event.ability.guid === SPELLS.HIGH_TIDE_BUFF.id) {
```

File: src/parser/shaman/restoration/modules/talents/HighTide.js [hunk 3]
```diff
@@ -21,22 +21,65 @@ const bounceReduction = 0.7;
...
+      this.unusedHighTides += this.currentHighTideBuff; //if there were leftovers when buff expires
+      this.currentHighTideBuff = 0;
+    }
+  }
+
+  // on a chain heal cast and buff counter is greater than 0, adds one to used counter 
+  // and reduces buff counter by one.
+  onChainHealCast(event) {
+    if(event.ability.name === "Chain Heal" && this.currentHighTideBuff > 0) {
```
Comment (niseko) [inline src/parser/shaman/restoration/modules/talents/HighTide.js]:
```suggestion
    if(this.currentHighTideBuff > 0) {
```
It's already guaranteed to be Chain Heal by how you've set up the event listener.

File: src/parser/shaman/restoration/modules/talents/HighTide.js [hunk 4]
```diff
@@ -21,22 +21,65 @@ const bounceReduction = 0.7;
...
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(SPELLS.CHAIN_HEAL), this.onChainHealCast);
+    this.addEventListener(Events.applybuff, this.onHighTideBuff);
+    this.addEventListener(Events.applybuffstack, this.onHighTideBuff); //possible to get a re-application of the buff
+    this.addEventListener(Events.removebuff, this.onHighTideRemoveBuff);
+  }
+
+  // in the event of a High Tide buff, if a buff still exists adds it to unused then resets buff counter to 2
+  onHighTideBuff(event) {
+    if(event.ability.guid === 288675) {
+      this.unusedHighTides += this.currentHighTideBuff; //not sure if this is possible
```
Comment (niseko) [inline src/parser/shaman/restoration/modules/talents/HighTide.js]: it is possible to overwrite your buff with a new one, yes

File: src/parser/shaman/restoration/CHANGELOG.js [hunk 5]
```diff
@@ -1,11 +1,12 @@
 import React from 'react';
 
-import { niseko, blazyb } from 'CONTRIBUTORS';
+import { niseko, blazyb, Torothin } from 'CONTRIBUTORS';
 import SPELLS from 'common/SPELLS';
 import SpellLink from 'common/SpellLink';
 import { change, date } from 'common/changelog';
 
 export default [
+  change(date(2020, 2, 28), <><SpellLink id={SPELLS.HIGH_TIDE_TALENT.id} /> buff usage added as a value tooltip for Hight Tide healing.</>, [Torothin]),
```
Comment (niseko) [inline src/parser/shaman/restoration/CHANGELOG.js]:
```suggestion
  change(date(2020, 3, 16), <><SpellLink id={SPELLS.HIGH_TIDE_TALENT.id} /> buff usage added as a value tooltip for Hight Tide healing.</>, [Torothin]),
```

File: src/parser/shaman/restoration/modules/talents/HighTide.js (hunk 1 above)
Comment (torothin) [reply src/parser/shaman/restoration/modules/talents/HighTide.js]: Cleaned this up to do spell checking using .spell method at the event listener and removed if statement from this method.

File: src/parser/shaman/restoration/modules/talents/HighTide.js (hunk 3 above)
Comment (torothin) [reply src/parser/shaman/restoration/modules/talents/HighTide.js]: fixed

File: src/parser/shaman/restoration/CHANGELOG.js (hunk 5 above)
Comment (torothin) [reply src/parser/shaman/restoration/CHANGELOG.js]: fixed

File: src/parser/shaman/restoration/modules/talents/HighTide.js (hunk 4 above)
Comment (torothin) [reply src/parser/shaman/restoration/modules/talents/HighTide.js]: Cleaned this up to do spell checking using .spell method at the event listener and removed if statement from this method.

File: src/parser/shaman/restoration/modules/talents/HighTide.js (hunk 2 above)
Comment (torothin) [reply src/parser/shaman/restoration/modules/talents/HighTide.js]: Cleaned this up to do spell checking using .spell method at the event listener and removed if statement from this method.

**Pull request merged**